keps/sig-api-machinery/2309-idl/syntax.md

```diff
@@ -0,0 +1,635 @@
+- [Formal(-ish) Grammar](#formal-ish-grammar)
+<!-- /toc -->
+
+## Design Goals
+
+1. High-level concepts and primitives in kubernetes should have
+   corresponding concepts in the language.  For instance:
+
+   * Quantity is fundamentally a primitive in Kubernetes, and thus should
+     have a corresponding primitive in our IDL.
+
+   * Kind is a fundamental concept in Kubernetes, so it should just be
+     possible to declare a kind in the IDL.
+
+2. Kubernetes API requirements should be encoded in the language:
+   disallowed constructs (and many bad practices) in Kubernetes APIs
+   should not be possible to represent in the IDL.
+
+3. Kubernetes API recommendations should be the default behavior in the
+   language: when not explicitly overridden, users should get our
+   recommended behavior by default:
+
+   * For example, the recommendation is new list-maps have a merge key of
+     name.  When merge keys aren’t manually specified, this is the default
+     one.
+
+4. All information conveyed should be as structured as possible
+
+5. Naming in the IDL describes the serialized form.  Language bindings
+   (Go) are derived automatically from that.  Specific overrides for
+   legacy codebases occur via markers, since they are exceptions to the
+   rules.
+
+## Examples
+
+***Author's note***: the cannonical source for these examples is [in the
+repository][repo-example]; the example reproduced here is just to make it
+easier to comment.
+
+[repo-example]: https://github.com/DirectXMan12/idl/blob/main/kdlc/testdata/all-around.kdl
+
+```
+// double slashes add comments
+/* same with block comments */
+
+// Triple-slash comments denote documentation
+// If a line contains `# Words`, the following lines
+// will be placed in the corresponding OpenAPI documentation
+// field.
+
+/// core/v1 describes core k8s concepts
+group-version(group:"core", version:"v1") {
+    /// Pod is a group of related, colocated processes
+    ///
+    /// # Example
+    /// {"apiVersion": "v1", "kind": "Pod", "spec": {}}
+    kind Pod {
+        // types may be nested -- this is just a convenience for namespacing
+        // it does not imply privacy or anything
+
+        spec: Spec,
+        struct Spec {
+            // list-maps are ordered maps whose keys appear in their body.
+            // not specifying a key defaults to `key: name`.
+            // Go represents them as lists with specific merge keys, other
+            // languages may represent them other ways.
+            // They are automatically merged as maps in SMD.
+
+            // optional fields must list the keyword `optional` before the
+            // type.
+
+            /// list of volumes that can be mounted by containers belonging to
+            /// the pod
+            volumes: optional list-map(value: Volume),
+
+            // ...
+
+            dnsPolicy: optional(default: ClusterFirst) DNSPolicy,
+
+            // simple-maps are unordered maps from string-type values to
```

Comment:
s/simple-map/dict ?  I think "simple" is a historical artifact and not something users should think about.

Reply:
fair.  I wanted to make clear the distiction between "list-map" & "simple-map", but if y'all think "dict" is clear enough, that's reasonable to me :-)

Reply:
I have a slight problem with the appearance of new concepts like simple-map. Or in other words: I am surprising there isn't more alignment with server-side-apply schemas. Shouldn't we expect all of https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/555-server-side-apply#api-topology to show-up here 1:1? Or are SSA concepts lower-level and everything here will be translated? If yes, such a translation would be helpful to be written out.

cc @apelisse

Reply:
SSA concepts strike me as kind of the slightly wrong granularity.  I tried to lay out my thoughts in the comments on the grammar, but I can copy here again:

lists are basically 1-to-1 with SSA:

- list --> list of type atomic
- list-map --> list of type map (naming is very intentional to suggest "this is how you do a map of structs in Kubernetes")
- set --> list of type set

Each of these is a different "conceptual" data structure IMO -- they just happen to have similar serialization details in JSON & proto.  They might have different representations in a programming language (many languages have native types for sets, for instance, or might want a helper that makes it easy to look up list-map items by key)

simple-map is a "map" in SSA.  It's intentionally `simple-map` and not just `map` because folks tend to reach for `map` when they actually want list-map, according to our API conventions.  I've seen this repeatedly.  By calling it `list-map` and `simple-map`, we make more obvious the distinction and give it a mnemonic to remember it by -- `simple-map` is for selectors and such (maps of "simple" values like primitives and on case of a list) vs `list-maps` for more complicated structures.

The lack of "atomic" version of maps is an oversight, but probably best addressed by the addition of an option to simple-map (e.g. `simple-map(atomic: true)`).

Does that make sense?